api/logic/modplatform/atlauncher/ATLPackInstallTask.cpp

```diff
@@ -0,0 +1,194 @@
+    {
+        emitFailed(tr("Unknown loader type: ") + m_version.loader.type);
+        return;
+    }
+    components->saveNow();
+
+    jobPtr.reset(new NetJob(tr("Mod download")));
+    QStringList jarmods;
+    for(const auto& mod : m_version.mods) {
+        QString relpath;
+        switch(mod.type) {
+            case ModType::Forge:
+                // todo: detect Forge version and install through a proper component
+            case ModType::Jar:
+                jarmods.push_back(mod.file);
+                relpath = "jarmods";
+                break;
+            case ModType::Mods:
+                relpath = FS::PathCombine("minecraft", "mods");
+                break;
+            case ModType::Flan:
+                relpath = FS::PathCombine("minecraft", "Flan");
+                break;
+            case ModType::Dependency:
+                relpath = FS::PathCombine("minecraft", "mods", m_version.pack.minecraft);
+                break;
+            case ModType::Ic2Lib:
+                relpath = FS::PathCombine("minecraft", "mods", "ic2");
+                break;
+            case ModType::DenLib:
+                relpath = FS::PathCombine("minecraft", "mods", "denlib");
+                break;
+            case ModType::Coremods:
+                relpath = FS::PathCombine("minecraft", "coremods");
+                break;
+            case ModType::MCPC:
+                // we can safely ignore MCPC server jar
+                break;
+            case ModType::Plugins:
+                relpath = FS::PathCombine("minecraft", "plugins");
+                break;
+            case ModType::Extract:
+            case ModType::Decomp:
+                // todo(merged): fail hard
+                qWarning() << "Unsupported mod type: " + mod.type_raw;
+                continue;
+            case ModType::ResourcePack:
+                relpath = FS::PathCombine("minecraft", "resourcepacks");
+                break;
+            case ModType::Unknown:
+                emitFailed(tr("Unknown mod type: ") + mod.type_raw);
+                return;
+        }
+        auto path = FS::PathCombine(m_stagingPath, relpath, mod.file);
+
+        QString url;
+        switch(mod.download) {
+            case DownloadType::Server:
+                url = BuildConfig.ATL_DOWNLOAD_SERVER + mod.url;
+                break;
+            case DownloadType::Browser:
+                emitFailed(tr("Unsupported download type: ") + mod.download_raw);
+                return;
+            case DownloadType::Direct:
+                url = mod.url;
+                break;
+            case DownloadType::Unknown:
+                emitFailed(tr("Unknown download type: ") + mod.download_raw);
+                return;
+        }
+
+        qDebug() << "Will download" << url << "to" << path;
+        auto dl = Net::Download::makeFile(url, path);
+        jobPtr->addNetAction(dl);
+    }
+    components->installJarMods(jarmods);
```

Comment:
This doesn't work at the moment, it needs to be run after the jarmods exist - unfortunately my attempt to do
```cpp
    connect(jobPtr.get(), &NetJob::succeeded, this, [&]()
    {
        jobPtr.reset();
        components->installJarMods(jarmods);
        instanceSettings->resumeSave();
        emitSucceeded();
    });
```
was unsuccessful - I welcome input on how to handle this *nicely*.

Reply:
Try getComponents()->saveNow() after installing the mods.